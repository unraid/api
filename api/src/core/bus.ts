import NanoBus from 'nanobus';

/**
 * Graphql event bus.
 */
export const bus = new NanoBus();
