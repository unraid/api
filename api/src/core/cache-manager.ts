/*!
 * Copyright 2019-2022 Lime Technology Inc. All rights reserved.
 * Written by: Alexis Tyler
 */

import { Cache } from 'clean-cache';

export const caches = new Map<string, Cache>();

// In seconds
const ONE_MINUTE = 60 * 1_000;

/**
 * Cache manager.
 */
export class CacheManager {
	ttl: number;

	constructor(private readonly name: string, ttl: number = ONE_MINUTE) {
		// Get cache
		let cache = caches.get(name);

		this.ttl = ttl;

		// Create new cache if we can't find one
		if (!cache) {
			cache = new Cache(ttl);
			caches.set(name, cache);
		}
	}

	get<T>(key: string): T {
		return caches.get(this.name)?.get(key);
	}

	set<T = undefined>(key: string, value: T, ttl?: number): T {
		// Get cache
		const cache = caches.get(this.name);

		// Check for existing entry and return that
		const item: T = cache?.get(key);
		if (item) {
			return item;
		}

		// Update cache
		// @ts-expect-error Todo fix the value type here when refactored to global store
		cache?.add(key, value, ttl ?? this.ttl ?? ONE_MINUTE);
		return value;
	}

	keys(): string[] {
		return [...caches.keys()];
	}
}
