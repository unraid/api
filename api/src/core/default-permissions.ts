export interface Permission { resource: string, action: string, attributes: string }
export interface Role {
	permissions: Array<Permission>
	extends?: string;
}

export const admin: Role = {
    extends: 'user',
    permissions: [
        // @NOTE: Uncomment the first line to enable creation of api keys.
        //        See the README.md for more information.
        // @WARNING: This is currently unsupported, please be careful.
        // { resource: 'apikey', action: 'create:any', attributes: '*' },
        { resource: 'apikey', action: 'read:any', attributes: '*' },
        { resource: 'array', action: 'read:any', attributes: '*' },
        { resource: 'cpu', action: 'read:any', attributes: '*' },
        {
            resource: 'crash-reporting-enabled',
            action: 'read:any',
            attributes: '*',
        },
        { resource: 'device', action: 'read:any', attributes: '*' },
        { resource: 'device/unassigned', action: 'read:any', attributes: '*' },
        { resource: 'disk', action: 'read:any', attributes: '*' },
        { resource: 'disk/settings', action: 'read:any', attributes: '*' },
        { resource: 'display', action: 'read:any', attributes: '*' },
        { resource: 'docker/container', action: 'read:any', attributes: '*' },
        { resource: 'docker/network', action: 'read:any', attributes: '*' },
        { resource: 'flash', action: 'read:any', attributes: '*' },
        { resource: 'info', action: 'read:any', attributes: '*' },
        { resource: 'license-key', action: 'read:any', attributes: '*' },
        { resource: 'machine-id', action: 'read:any', attributes: '*' },
        { resource: 'memory', action: 'read:any', attributes: '*' },
        { resource: 'notifications', action: 'read:any', attributes: '*' },
        { resource: 'online', action: 'read:any', attributes: '*' },
        { resource: 'os', action: 'read:any', attributes: '*' },
        { resource: 'owner', action: 'read:any', attributes: '*' },
        { resource: 'parity-history', action: 'read:any', attributes: '*' },
        { resource: 'permission', action: 'read:any', attributes: '*' },
        { resource: 'registration', action: 'read:any', attributes: '*' },
        { resource: 'servers', action: 'read:any', attributes: '*' },
        { resource: 'service', action: 'read:any', attributes: '*' },
        { resource: 'service/emhttpd', action: 'read:any', attributes: '*' },
        { resource: 'service/unraid-api', action: 'read:any', attributes: '*' },
        { resource: 'services', action: 'read:any', attributes: '*' },
        { resource: 'share', action: 'read:any', attributes: '*' },
        { resource: 'software-versions', action: 'read:any', attributes: '*' },
        { resource: 'unraid-version', action: 'read:any', attributes: '*' },
        { resource: 'uptime', action: 'read:any', attributes: '*' },
        { resource: 'user', action: 'read:any', attributes: '*' },
        { resource: 'vars', action: 'read:any', attributes: '*' },
        { resource: 'vms', action: 'read:any', attributes: '*' },
        { resource: 'vms/domain', action: 'read:any', attributes: '*' },
        { resource: 'vms/network', action: 'read:any', attributes: '*' },
    ],
};

export const user: Role = {
	extends: 'guest',
	permissions: [
		{ resource: 'apikey', action: 'read:own', attributes: '*' },
		{ resource: 'permission', action: 'read:any', attributes: '*' },
	],
};

export const upc: Role = {
	extends: 'guest',
	permissions: [
		{ resource: 'apikey', action: 'read:own', attributes: '*' },
		{ resource: 'cloud', action: 'read:own', attributes: '*' },
		{ resource: 'config', action: 'read:any', attributes: '*' },
		{ resource: 'crash-reporting-enabled', action: 'read:any', attributes: '*' },
		{ resource: 'disk', action: 'read:any', attributes: '*' },
		{ resource: 'display', action: 'read:any', attributes: '*' },
		{ resource: 'flash', action: 'read:any', attributes: '*' },
		{ resource: 'os', action: 'read:any', attributes: '*' },
		{ resource: 'owner', action: 'read:any', attributes: '*' },
		{ resource: 'permission', action: 'read:any', attributes: '*' },
		{ resource: 'registration', action: 'read:any', attributes: '*' },
		{ resource: 'servers', action: 'read:any', attributes: '*' },
		{ resource: 'vars', action: 'read:any', attributes: '*' },
	],
};

export const my_servers: Role = {
    extends: 'guest',
    permissions: [
        { resource: 'dashboard', action: 'read:any', attributes: '*' },
        { resource: 'two-factor', action: 'read:own', attributes: '*' },
        { resource: 'array', action: 'read:any', attributes: '*' },
        { resource: 'docker/container', action: 'read:any', attributes: '*' },
        { resource: 'docker/network', action: 'read:any', attributes: '*' },
        { resource: 'notifications', action: 'read:any', attributes: '*' },
        { resource: 'vms/domain', action: 'read:any', attributes: '*' },
        { resource: 'unraid-version', action: 'read:any', attributes: '*' },
    ],
};

export const notifier: Role = {
	extends: 'guest',
	permissions: [
		{ resource: 'notifications', action: 'create:own', attributes: '*' },
	],
};

export const guest: Role = {
	permissions: [
		{ resource: 'me', action: 'read:any', attributes: '*' },
		{ resource: 'welcome', action: 'read:any', attributes: '*' },
	],
};

export const permissions: Record<string, Role> = {
	guest,
	user,
	admin,
	upc,
	my_servers,
	notifier,
};
