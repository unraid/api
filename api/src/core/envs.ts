/*!
 * Copyright 2019-2022 Lime Technology Inc. All rights reserved.
 * Written by: Alexis Tyler
 */

const _ = process.env;

/**
 * Proxy for process.env
 *
 * @note Add known envs here for better typing
 */
export const envs = {
	..._,
	NODE_ENV: _.NODE_ENV!,
	DEBUG: _.DEBUG === 'true',
	PORT: _.PORT!,
	NODE_API_PORT: _.NODE_API_PORT!,
	DRY_RUN: Boolean(_.DRY_RUN),
};
