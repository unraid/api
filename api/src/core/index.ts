export * as modules from '@app/core/modules';
export * as notifiers from '@app/core/notifiers';
export * as utils from '@app/core/utils';
export * from '@app/core/log';
export * from '@app/core/permission-manager';
export * from '@app/core/permissions';
export * from '@app/core/pubsub';
