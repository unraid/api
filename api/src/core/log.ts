import { pino } from 'pino';
import { LOG_TYPE } from '@app/environment';

import pretty from 'pino-pretty';

export const levels = [
    'trace',
    'debug',
    'info',
    'warn',
    'error',
    'fatal',
] as const;

export type LogLevel = typeof levels[number];

const level =
    levels[
        levels.indexOf(
            process.env.LOG_LEVEL?.toLowerCase() as (typeof levels)[number]
        )
    ] ?? 'info';

export const logDestination = pino.destination({
    sync: true,
});

const stream =
    LOG_TYPE === 'pretty'
        ? pretty({
              singleLine: true,
              hideObject: false,
              colorize: true,
              ignore: 'time,hostname,pid',
              destination: logDestination,
          })
        : logDestination;

export const logger = pino(
    {
        level,
        timestamp: () => `,"time":"${new Date().toISOString()}"`,
        formatters: {
            level: (label: string) => ({ level: label }),
        },
    },
    stream
);

export const internalLogger = logger.child({ logger: 'internal' });
export const appLogger = logger.child({ logger: 'app' });
export const mothershipLogger = logger.child({ logger: 'mothership' });
export const dashboardLogger = logger.child({ logger: 'dashboard' });
export const emhttpLogger = logger.child({ logger: 'emhttp' });
export const libvirtLogger = logger.child({ logger: 'libvirt' });
export const graphqlLogger = logger.child({ logger: 'graphql' });
export const dockerLogger = logger.child({ logger: 'docker' });
export const cliLogger = logger.child({ logger: 'cli' });
export const minigraphLogger = logger.child({ logger: 'minigraph' });
export const cloudConnectorLogger = logger.child({ logger: 'cloud-connector' });
export const upnpLogger = logger.child({ logger: 'upnp' });
export const keyServerLogger = logger.child({ logger: 'key-server' });
export const remoteAccessLogger = logger.child({ logger: 'remote-access' });
export const remoteQueryLogger = logger.child({ logger: 'remote-query' });
export const apiLogger = logger.child({ logger: 'api' });

export const loggers = [
    internalLogger,
    appLogger,
    mothershipLogger,
    dashboardLogger,
    emhttpLogger,
    libvirtLogger,
    graphqlLogger,
    dockerLogger,
    cliLogger,
    minigraphLogger,
    cloudConnectorLogger,
    upnpLogger,
    keyServerLogger,
    remoteAccessLogger,
    remoteQueryLogger,
    apiLogger,
];
