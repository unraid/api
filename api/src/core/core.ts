/*!
 * Copyright 2019-2022 Lime Technology Inc. All rights reserved.
 * Written by: Alexis Tyler
 */

import { logger, nchanLogger } from '@app/core/log';
import { subscribeToNchanEndpoint } from '@app/core/utils';
import { config } from '@app/core/config';
import * as watchers from '@app/core/watchers';

/**
 * Decorated loading logger.
 * @param namespace
 */
const loadingLogger = (namespace: string): void => {
	logger.debug('Loading %s', namespace);
};

/**
 * Start all watchers.
 */
const loadWatchers = async (): Promise<void> => {
	if (config.safeMode) {
		logger.debug('Skipping loading watchers');
		return;
	}

	loadingLogger('watchers');

	// Start each watcher
	Object.entries(watchers).forEach(([name, watcher]) => {
		logger.debug('Loading %s watcher', name);
		watcher().start();
	});
};

/**
 * Connect to nchan endpoints.
 *
 * @param endpoints
 */
const connectToNchanEndpoints = async (endpoints: string[]): Promise<void> => {
	nchanLogger.debug('Connected, setting-up endpoints.');
	const connections = endpoints.map(async endpoint => subscribeToNchanEndpoint(endpoint));
	await Promise.all(connections);
};

/**
 * Start nchan subscriptions
 *
 * @name core.loadNchan
 * @async
 * @private
 */
const loadNchan = async (): Promise<void> => {
	const endpoints = ['devs', 'disks', 'sec', 'sec_nfs', 'shares', 'users', 'var'];

	logger.debug('Trying to connect to nchan');

	// Connect to each known endpoint
	await connectToNchanEndpoints(endpoints);
};

/**
 * Core loaders.
 */
const loaders = {
	watchers: loadWatchers,
};

/**
 * Main load function
 *
 * @name core.load
 */
const load = async (): Promise<void> => {
	await loadWatchers();
};

export const core = {
	loaders,
	load,
	loadNchan,
};
