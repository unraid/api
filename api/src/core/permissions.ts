import { apiLogger } from '@app/core/log';
import { RolesBuilder } from 'nest-access-control';

export interface Permission {
    role?: string;
    resource: string;
    action: string;
    attributes: string;
}
export interface Role {
    permissions: Array<Permission>;
    extends?: string;
}

// Use built in permissions
const roles: Record<string, Role> = {
    admin: {
        extends: 'guest',
        permissions: [
            { resource: 'apikey', action: 'read:any', attributes: '*' },
            { resource: 'cloud', action: 'read:own', attributes: '*' },
            { resource: 'config', action: 'update:own', attributes: '*' },
            { resource: 'config', action: 'read:any', attributes: '*' },
            { resource: 'connect', action: 'read:own', attributes: '*' },
            { resource: 'connect', action: 'update:own', attributes: '*' },
            { resource: 'customizations', action: 'read:any', attributes: '*' },
            { resource: 'array', action: 'read:any', attributes: '*' },
            { resource: 'cpu', action: 'read:any', attributes: '*' },
            {
                resource: 'crash-reporting-enabled',
                action: 'read:any',
                attributes: '*',
            },
            { resource: 'device', action: 'read:any', attributes: '*' },
            {
                resource: 'device/unassigned',
                action: 'read:any',
                attributes: '*',
            },
            { resource: 'disk', action: 'read:any', attributes: '*' },
            { resource: 'disk/settings', action: 'read:any', attributes: '*' },
            { resource: 'display', action: 'read:any', attributes: '*' },
            {
                resource: 'docker/container',
                action: 'read:any',
                attributes: '*',
            },
            { resource: 'docker/network', action: 'read:any', attributes: '*' },
            { resource: 'flash', action: 'read:any', attributes: '*' },
            { resource: 'info', action: 'read:any', attributes: '*' },
            { resource: 'license-key', action: 'read:any', attributes: '*' },
            { resource: 'logs', action: 'read:any', attributes: '*' },
            { resource: 'machine-id', action: 'read:any', attributes: '*' },
            { resource: 'memory', action: 'read:any', attributes: '*' },
            { resource: 'notifications', action: 'read:any', attributes: '*' },
            {
                resource: 'notifications',
                action: 'create:any',
                attributes: '*',
            },
            { resource: 'online', action: 'read:any', attributes: '*' },
            { resource: 'os', action: 'read:any', attributes: '*' },
            { resource: 'owner', action: 'read:any', attributes: '*' },
            { resource: 'parity-history', action: 'read:any', attributes: '*' },
            { resource: 'permission', action: 'read:any', attributes: '*' },
            { resource: 'registration', action: 'read:any', attributes: '*' },
            { resource: 'servers', action: 'read:any', attributes: '*' },
            { resource: 'service', action: 'read:any', attributes: '*' },
            {
                resource: 'service/emhttpd',
                action: 'read:any',
                attributes: '*',
            },
            {
                resource: 'service/unraid-api',
                action: 'read:any',
                attributes: '*',
            },
            { resource: 'services', action: 'read:any', attributes: '*' },
            { resource: 'share', action: 'read:any', attributes: '*' },
            {
                resource: 'software-versions',
                action: 'read:any',
                attributes: '*',
            },
            { resource: 'unraid-version', action: 'read:any', attributes: '*' },
            { resource: 'uptime', action: 'read:any', attributes: '*' },
            { resource: 'user', action: 'read:any', attributes: '*' },
            { resource: 'vars', action: 'read:any', attributes: '*' },
            { resource: 'vms', action: 'read:any', attributes: '*' },
            { resource: 'vms/domain', action: 'read:any', attributes: '*' },
            { resource: 'vms/network', action: 'read:any', attributes: '*' },
        ],
    },
    upc: {
        extends: 'guest',
        permissions: [
            { resource: 'apikey', action: 'read:own', attributes: '*' },
            { resource: 'cloud', action: 'read:own', attributes: '*' },
            { resource: 'config', action: 'read:any', attributes: '*' },
            {
                resource: 'crash-reporting-enabled',
                action: 'read:any',
                attributes: '*',
            },
            { resource: 'customizations', action: 'read:any', attributes: '*' },
            { resource: 'disk', action: 'read:any', attributes: '*' },
            { resource: 'display', action: 'read:any', attributes: '*' },
            { resource: 'flash', action: 'read:any', attributes: '*' },
            { resource: 'info', action: 'read:any', attributes: '*' },
            { resource: 'logs', action: 'read:any', attributes: '*' },
            { resource: 'os', action: 'read:any', attributes: '*' },
            { resource: 'owner', action: 'read:any', attributes: '*' },
            { resource: 'permission', action: 'read:any', attributes: '*' },
            { resource: 'registration', action: 'read:any', attributes: '*' },
            { resource: 'servers', action: 'read:any', attributes: '*' },
            { resource: 'vars', action: 'read:any', attributes: '*' },
            { resource: 'config', action: 'update:own', attributes: '*' },
            { resource: 'connect', action: 'read:own', attributes: '*' },
            { resource: 'connect', action: 'update:own', attributes: '*' },
            { resource: 'notifications', action: 'read:any', attributes: '*' },
            { resource: 'notifications', action: 'update:any', attributes: '*' },
        ],
    },
    my_servers: {
        extends: 'guest',
        permissions: [
            { resource: 'array', action: 'read:any', attributes: '*' },
            { resource: 'config', action: 'read:any', attributes: '*' },
            { resource: 'connect', action: 'read:any', attributes: '*' },
            {
                resource: 'connect/dynamic-remote-access',
                action: 'read:any',
                attributes: '*',
            },
            {
                resource: 'connect/dynamic-remote-access',
                action: 'update:own',
                attributes: '*',
            },
            { resource: 'customizations', action: 'read:any', attributes: '*' },
            { resource: 'dashboard', action: 'read:any', attributes: '*' },
            { resource: 'display', action: 'read:any', attributes: '*' },
            {
                resource: 'docker/container',
                action: 'read:any',
                attributes: '*',
            },
            { resource: 'docker', action: 'read:any', attributes: '*' },
            {
                resource: 'docker/container',
                action: 'read:any',
                attributes: '*',
            },
            { resource: 'info', action: 'read:any', attributes: '*' },
            { resource: 'logs', action: 'read:any', attributes: '*' },
            { resource: 'network', action: 'read:any', attributes: '*' },
            { resource: 'notifications', action: 'read:any', attributes: '*' },
            { resource: 'services', action: 'read:any', attributes: '*' },
            { resource: 'vars', action: 'read:any', attributes: '*' },
            { resource: 'vms', action: 'read:any', attributes: '*' },
            { resource: 'vms/domain', action: 'read:any', attributes: '*' },
            { resource: 'unraid-version', action: 'read:any', attributes: '*' },
        ],
    },
    notifier: {
        extends: 'guest',
        permissions: [
            {
                resource: 'notifications',
                action: 'create:own',
                attributes: '*',
            },
        ],
    },
    guest: {
        permissions: [
            { resource: 'me', action: 'read:any', attributes: '*' },
            { resource: 'welcome', action: 'read:any', attributes: '*' },
        ],
    },
};

export const setupPermissions = (): RolesBuilder => {
    // First create an array of permissions that will be used as the base permission set for the app
    const grantList = Object.entries(roles).reduce<Array<Permission>>(
        (acc, [roleName, role]) => {
            if (role.permissions) {
                role.permissions.forEach((permission) => {
                    acc.push({
                        ...permission,
                        role: roleName,
                    });
                });
            }
            return acc;
        },
        []
    );

    const ac = new RolesBuilder(grantList);

    // Next, Extend roles
    Object.entries(roles).forEach(([roleName, role]) => {
        if (role.extends) {
            ac.extendRole(roleName, role.extends);
        }
    });

    apiLogger.debug('Possible Roles: %o', ac.getRoles());

    return ac;
};

export const ac = null;
