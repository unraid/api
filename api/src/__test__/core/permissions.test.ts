import { expect, test } from 'vitest';
import { setupPermissions } from '@app/core/permissions';

test('Returns default permissions', () => {
	expect(setupPermissions()).toMatchSnapshot();
});
