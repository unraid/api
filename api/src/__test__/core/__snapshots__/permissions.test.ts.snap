// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Returns default permissions 1`] = `
RolesBuilder {
  "_grants": {
    "admin": {
      "$extend": [
        "guest",
      ],
      "apikey": {
        "read:any": [
          "*",
        ],
      },
      "array": {
        "read:any": [
          "*",
        ],
      },
      "cloud": {
        "read:own": [
          "*",
        ],
      },
      "config": {
        "update:own": [
          "*",
        ],
      },
      "connect": {
        "read:own": [
          "*",
        ],
        "update:own": [
          "*",
        ],
      },
      "cpu": {
        "read:any": [
          "*",
        ],
      },
      "crash-reporting-enabled": {
        "read:any": [
          "*",
        ],
      },
      "customizations": {
        "read:any": [
          "*",
        ],
      },
      "device": {
        "read:any": [
          "*",
        ],
      },
      "device/unassigned": {
        "read:any": [
          "*",
        ],
      },
      "disk": {
        "read:any": [
          "*",
        ],
      },
      "disk/settings": {
        "read:any": [
          "*",
        ],
      },
      "display": {
        "read:any": [
          "*",
        ],
      },
      "docker/container": {
        "read:any": [
          "*",
        ],
      },
      "docker/network": {
        "read:any": [
          "*",
        ],
      },
      "flash": {
        "read:any": [
          "*",
        ],
      },
      "info": {
        "read:any": [
          "*",
        ],
      },
      "license-key": {
        "read:any": [
          "*",
        ],
      },
      "logs": {
        "read:any": [
          "*",
        ],
      },
      "machine-id": {
        "read:any": [
          "*",
        ],
      },
      "memory": {
        "read:any": [
          "*",
        ],
      },
      "notifications": {
        "create:any": [
          "*",
        ],
        "read:any": [
          "*",
        ],
      },
      "online": {
        "read:any": [
          "*",
        ],
      },
      "os": {
        "read:any": [
          "*",
        ],
      },
      "owner": {
        "read:any": [
          "*",
        ],
      },
      "parity-history": {
        "read:any": [
          "*",
        ],
      },
      "permission": {
        "read:any": [
          "*",
        ],
      },
      "registration": {
        "read:any": [
          "*",
        ],
      },
      "servers": {
        "read:any": [
          "*",
        ],
      },
      "service": {
        "read:any": [
          "*",
        ],
      },
      "service/emhttpd": {
        "read:any": [
          "*",
        ],
      },
      "service/unraid-api": {
        "read:any": [
          "*",
        ],
      },
      "services": {
        "read:any": [
          "*",
        ],
      },
      "share": {
        "read:any": [
          "*",
        ],
      },
      "software-versions": {
        "read:any": [
          "*",
        ],
      },
      "unraid-version": {
        "read:any": [
          "*",
        ],
      },
      "uptime": {
        "read:any": [
          "*",
        ],
      },
      "user": {
        "read:any": [
          "*",
        ],
      },
      "vars": {
        "read:any": [
          "*",
        ],
      },
      "vms": {
        "read:any": [
          "*",
        ],
      },
      "vms/domain": {
        "read:any": [
          "*",
        ],
      },
      "vms/network": {
        "read:any": [
          "*",
        ],
      },
    },
    "guest": {
      "me": {
        "read:any": [
          "*",
        ],
      },
      "welcome": {
        "read:any": [
          "*",
        ],
      },
    },
    "my_servers": {
      "$extend": [
        "guest",
      ],
      "array": {
        "read:any": [
          "*",
        ],
      },
      "config": {
        "read:any": [
          "*",
        ],
      },
      "connect": {
        "read:any": [
          "*",
        ],
      },
      "connect/dynamic-remote-access": {
        "read:any": [
          "*",
        ],
        "update:own": [
          "*",
        ],
      },
      "customizations": {
        "read:any": [
          "*",
        ],
      },
      "dashboard": {
        "read:any": [
          "*",
        ],
      },
      "display": {
        "read:any": [
          "*",
        ],
      },
      "docker": {
        "read:any": [
          "*",
        ],
      },
      "docker/container": {
        "read:any": [
          "*",
        ],
      },
      "info": {
        "read:any": [
          "*",
        ],
      },
      "logs": {
        "read:any": [
          "*",
        ],
      },
      "network": {
        "read:any": [
          "*",
        ],
      },
      "notifications": {
        "read:any": [
          "*",
        ],
      },
      "services": {
        "read:any": [
          "*",
        ],
      },
      "unraid-version": {
        "read:any": [
          "*",
        ],
      },
      "vars": {
        "read:any": [
          "*",
        ],
      },
      "vms": {
        "read:any": [
          "*",
        ],
      },
      "vms/domain": {
        "read:any": [
          "*",
        ],
      },
    },
    "notifier": {
      "$extend": [
        "guest",
      ],
      "notifications": {
        "create:own": [
          "*",
        ],
      },
    },
    "upc": {
      "$extend": [
        "guest",
      ],
      "apikey": {
        "read:own": [
          "*",
        ],
      },
      "cloud": {
        "read:own": [
          "*",
        ],
      },
      "config": {
        "read:any": [
          "*",
        ],
        "update:own": [
          "*",
        ],
      },
      "connect": {
        "read:own": [
          "*",
        ],
        "update:own": [
          "*",
        ],
      },
      "crash-reporting-enabled": {
        "read:any": [
          "*",
        ],
      },
      "customizations": {
        "read:any": [
          "*",
        ],
      },
      "disk": {
        "read:any": [
          "*",
        ],
      },
      "display": {
        "read:any": [
          "*",
        ],
      },
      "flash": {
        "read:any": [
          "*",
        ],
      },
      "info": {
        "read:any": [
          "*",
        ],
      },
      "logs": {
        "read:any": [
          "*",
        ],
      },
      "os": {
        "read:any": [
          "*",
        ],
      },
      "owner": {
        "read:any": [
          "*",
        ],
      },
      "permission": {
        "read:any": [
          "*",
        ],
      },
      "registration": {
        "read:any": [
          "*",
        ],
      },
      "servers": {
        "read:any": [
          "*",
        ],
      },
      "vars": {
        "read:any": [
          "*",
        ],
      },
    },
  },
  "_isLocked": false,
}
`;
